format/spec.md

```diff
@@ -951,24 +958,45 @@ Types are serialized according to this table:
+| **`timestamp`**    | **`json long`**   | `1646277378000000`               | Stores microseconds from 1970-01-01 00:00:00.000000                                                                                                                                                                                                                                                                       |
+| **`timestamptz`**  | **`json long`**   | `1646277378000000`               | Stores microseconds from 1970-01-01 00:00:00.000000 UTC                                                                                                                                                                                                                                                                   |
+| **`string`**       | **`string`**      | `"foo"`                          |                                                                                                                                                                                                                                                                                                                           |
+| **`uuid`**         | **`string`**      | `"eb26bdb1-a1d8-4aa6-990e-da940875492c"` | Stores the lowercase uuid string                                                                                                                                                                                                                                                                                          |
+| **`fixed(L)`**     | **`string`**      | `"0x3162"`                       | Stored as a hexadecimal byte literal string, prefixex by `0x`                                                                                                                                                                                                                                                             |
+| **`binary`**       | **`string`**      | `"0x3162"`                       | Stored as a hexadecimal byte literal string, prefixex by `0x`                                                                                                                                                                                                                                                             |
+| **`struct`**       | **`object`**      | `{"a": 1, "foo": "bar"}`         | Use a JSON map to represent struct data, the keys are the nested fields' names in the struct schema, and the values are value literals of corresponding fields' type                                                                                                                                                      |
```

Comment:
This requires a fairly complicated process to rewrite the default value when the schema changes because it uses field names instead of field IDs. Have you considered other representations, like a list of fields: `[ {"field-id": 5, "value": "s"}, ... ]`?

Reply:
After thinking about this for a couple days, my preference is to use the default values from the struct fields.

If an implementation doesn't update the default struct value then the most reasonable choice is to use the default that was required to create the field. That also means that we don't need to worry about traversing up the schema tree and updating default values when a nested struct changes.

My preference is to encode the default value that the user supplied and add the defaults of child fields to that value when fields are not present. For initial default values, we would use initial defaults. And for write defaults we would use write defaults.

This is too long to go in the JSON encoding section. This should be covered in the default values section above.

Reply:
> This requires a fairly complicated process to rewrite the default value when the schema changes because it uses field names instead of field IDs. Have you considered other representations, like a list of fields: `[ {"field-id": 5, "value": "s"}, ... ]`?

I think you are right about using field id to better track fields in terms of schema evolution. do you think array and map also need to be encoded this way (using field-id instead of name)? My feeling is they are unnecessary because array and map can't have schema evolution themselves anyway?

Reply:
> My preference is to encode the default value that the user supplied and add the defaults of child fields to that value when fields are not present. For initial default values, we would use initial defaults. And for write defaults we would use write defaults.

Are you saying if the a default value is missing for a outer struct, its child default value should be automatically propagated to its own level ?

I feel adding this propagation logic (either traverse up the tree, or down the tree) will be too complicated. I have a different idea, how about we treat all the default value as independent?

Let's say there is a struct called `foo` and its one child called `foo.bar`.

If the user is to query and project `foo`, i.e. `select foo`, we use the default value at the `foo` level.
If the user is to query and project `foo.bar`, i.e. `select foo.bar`, we use the default value at the `foo.bar` level.
Basically, I'm saying the default value of `foo` and `foo.bar` are independent, and users also set those 2 values independently.

Reply:
> @rdblue In the case of default values on files which have been name-mapped, would we still do name-mapping first and then use the field-id mapping correct? If so I think that sounds pretty reasonable although it would make things a bit more difficult to read in the JSON file.

I don't think that name mapping would have any effect here. Name mapping is a way to produce an Iceberg schema, but Iceberg schemas require field IDs. So everything should have an ID by this point.

If we were to use names, then the ID/name pairings would come from the schema itself -- hence the requirement that we update the defaults as the names change. But that seems needlessly difficult to implement so I think it makes much more sense to track default values by field ID.

Reply:
> do you think array and map also need to be encoded this way (using field-id instead of name)?

No, I think that map and array defaults should be stored as JSON lists. The element or key/value IDs are fairly easy to match up.

> Are you saying if the a default value is missing for a outer struct, its child default value should be automatically propagated to its own level ?

Yes. I think we have to do something because we need to define not only what to write, but how to read. It is likely that some implementation will fail to update a default struct when a child field is added. We can do a two things in that case: fail because we don't know how to handle the child field, or look at the child field and use its default value.

I doubt that we will be able to simply "require" keeping the default updated, and clearly refusing to read the table because of a default value issue is not a good idea. So that leaves using the child's default value.

If we choose to do this now, then we can simplify the requirements because you don't _have_ to update parent default values.

Reply:
> Basically, I'm saying the default value of foo and foo.bar are independent, and users also set and change those 2 values independently.

How is this possible? If I have `foo struct<required bap string> default struct("zap")` and I add `bar` to get `foo struct<required bap string, required bar int default 0>` then the defaults cannot be completely independent because `bar` is required. You either have to update the `foo` default to `struct("zap", 0)` when adding `bar` (which is not reliable), use the default for `bar`, or fail because `bar` is required but has not value in the struct's default value.